pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -1048,6 +1082,57 @@ func (m *ManagerImpl) GetDevices(podUID, containerName string) ResourceDeviceIns
 	return m.podDevices.getContainerDevices(podUID, containerName)
 }
 
+func (m *ManagerImpl) UpdateAllocatedResourcesStatus(pod *v1.Pod, status *v1.PodStatus) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	// Today we ignore edge cases that are not likely to happen:
+	//  - update statuses for containers that are in spec, but not in status
+	//  - update statuses for resources requested in spec, but with no information in podDevices
+	for i, containerStatus := range status.ContainerStatuses {
+		devices := m.podDevices.getContainerDevices(string(pod.UID), containerStatus.Name)
+
+		for resourceName, deviceInstances := range devices {
+			for id, d := range deviceInstances {
+				health := pluginapi.Healthy
+				// this is unlikely, but check for existence here anyways
+				if r, ok := m.allDevices[resourceName]; ok {
+					if _, ok := r[id]; ok {
+						health = m.allDevices[resourceName][id].Health
+					}
+				}
+
+				d.Health = health
+
+				deviceInstances[id] = d
+			}
+		}
+
+		for resourceName, dI := range devices {
+			resourceStatus := v1.ResourceStatus{
+				Name:      v1.ResourceName(resourceName),
+				Resources: []v1.ResourceHealth{},
+			}
+
+			for id, d := range dI {
+				health := v1.ResourceHealthStatusHealthy
+				if d.Health != pluginapi.Healthy {
+					health = v1.ResourceHealthStatusUnhealthy
+				}
+				resourceStatus.Resources = append(resourceStatus.Resources, v1.ResourceHealth{
+					ResourceID: v1.ResourceID(id),
+					Health:     health,
+				})
+			}
+
+			if status.ContainerStatuses[i].AllocatedResourcesStatus == nil {
+				status.ContainerStatuses[i].AllocatedResourcesStatus = []v1.ResourceStatus{}
+			}
+			status.ContainerStatuses[i].AllocatedResourcesStatus = append(status.ContainerStatuses[i].AllocatedResourcesStatus, resourceStatus)
```

Comment:
I think this infinitely appends? do we have a test which ensures we overwrite old statuses without creating an infinite list?
plus, do we want or need to sort the statuses?

Reply:
we can ensure we never infinitely append by adding an expectations on the tests, that the allocated resources size is constant and matches the expectations. Should be easy enough.

Reply:
good catch.